src/Workspaces/CSharp/Portable/Extensions/SemanticModelExtensions.cs

```diff
@@ -220,14 +220,40 @@ private static ImmutableArray<ParameterName> GenerateNames(IList<string> reserve
             // instead.
             var info = semanticModel.GetTypeInfo(expression, cancellationToken);
 
-            // If we can't determine the type, then fallback to some placeholders.
             var type = info.Type;
             var pluralize = Pluralize(semanticModel, type);
 
+            // We may be able to use the type's arguments to generate a name.
+            if (TryGenerateNameFromTypeArgument(type, capitalize, pluralize, out var typeArgumentParameterName))
+            {
+                return typeArgumentParameterName;
+            }
+
             var parameterName = type.CreateParameterName(capitalize);
             return pluralize ? parameterName.Pluralize() : parameterName;
         }
 
+        private static bool TryGenerateNameFromTypeArgument(ITypeSymbol type, bool capitalize, bool pluralize, out string parameterName)
+        {
+            var typeArguments = type.GetAllTypeArguments();
+
+            // We only consider generating a name if there's one type argument.
+            // This logic can potentially be expanded upon in the future.
+            if (typeArguments.Length == 1)
+            {
+                var typeArgument = typeArguments.Single().ToNameDisplayString();
+                if (SyntaxFacts.IsValidIdentifier(typeArgument))
+                {
+                    parameterName = capitalize ? typeArgument.ToPascalCase() : typeArgument.ToCamelCase();
+                    parameterName = pluralize ? parameterName.Pluralize() : parameterName;
+                    return true;
```

Comment:
i don't think this is a great approach.  say i have `IServiceLocator<Frob>`, this will generate the name `Frobs`, insead of `serviceLocator`.

I think what we want is to take the following approach if the type is generic *and* derives from IEnumerable<T>.  That way we could at least believe it's a collection type and thus we shlud pluralize the element ttype.